test: Add external test for pkcs11-provider

The-Mule opened the pull request:
This PR adds an external test for pkcs11-provider (https://github.com/latchset/pkcs11-provider). It runs the self-test of the provider.

##### Checklist
- [x] tests are added or updated

The-Mule commented:
Notice that there are currently 4 failures in the external test. All of them happens only when the (pkcs11-provider) self-test is executed through `./test/run_tests.pl` in OpenSSL. This test driver seems to be prepending "#" to the lines of openssl output and that is something the self-test does not really count with. Is there a way to change this somehow?

This is an example:

```
=================================== 54/60 ====================================
test:         pkcs11-provider:softokn / uri
start time:   16:38:03
duration:     0.11s
result:       exit status 1
command:      UBSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 TESTBLDDIR=/root/openssl/pkcs11-provider/builddir/tests ASAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1 MSAN_OPTIONS=halt_on_error=1:abort_on_error=1:print_summary=1:print_stacktrace=1 TEST_PATH=/root/openssl/pkcs11-provider/tests /root/openssl/pkcs11-provider/tests/test-wrapper uri-softokn.t
----------------------------------- stdout -----------------------------------
Executing /root/openssl/pkcs11-provider/tests/turi

## Check that storeutl returns URIs
 openssl storeutl -text pkcs11:
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%00;object=caCert;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%01;object=testCert;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%02;object=ecCert;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%03;object=ecPeerCert;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%05;object=testCert2;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%06;object=ecCert2;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%08;object=ecCert3;type=private
# URI pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%FD%41%3D%37%E0%FC%F2%D1%4D%86%7F%9C%EF%C6%33%6D;object=Fork-Test-fd413d37;type=private

## Check returned URIs work to find objects
$uri=# pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%00;object=caCert;type=private
 openssl storeutl -text "$uri"
----------------------------------- stderr -----------------------------------
Couldn't open file or uri # pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%00;object=caCert;type=private
80BBB690A37F0000:error:80000002:system library:file_open:No such file or directory:providers/implementations/storemgmt/file_store.c:263:calling stat(# pkcs11:model=NSS%203;manufacturer=Mozilla%20Foundation;serial=0000000000000000;token=NSS%20Certificate%20DB;id=%00%00;object=caCert;type=private)
80BBB690A37F0000:error:1608010C:STORE routines:inner_loader_fetch:unsupported:crypto/store/store_meth.c:360:No store loader found. For standard store loaders you need at least one of the default or base providers available. Did you forget to load them? Info: Global default library context, Scheme (# pkcs11 : 0), Properties (<null>)
==============================================================================
```

See the output of `openssl storeutl -text pkcs11:` above. This is the cause of each of the four failures.

beldmit commented:
@The-Mule try adding `export HARNESS_OSSL_PREFIX=''` to your shell script

slontis commented on test/README-external.md:
```diff
@@ -118,6 +118,35 @@ algorithms to be skipped, e.g. as follows:
+    $ ./config shared enable-external-tests
+    $ make
+
+pkcs11-provider requires meson and cmake for the build process. Moreover, it 
```
```suggestion
pkcs11-provider requires meson and cmake for the build process. Moreover, it
```

slontis commented:
I have not set up a external test before, but doesn't this require a ci.yml change?

levitte commented on test/README-external.md:
```diff
@@ -118,6 +118,35 @@ algorithms to be skipped, e.g. as follows:
+    $ ./config shared enable-external-tests
+    $ make
+
+pkcs11-provider requires meson and cmake for the build process. Moreover, it 
```
I don't see a trace of CMake files in https://github.com/latchset/pkcs11-provider.git

The-Mule replied:
Ah, you're right, I fixed that and rewrote the dependencies in distro-agnostic way too:

```
pkcs11-provider requires meson for the build process. Moreover, it requires
softhsm and nss softokn tokens and certtool, certutil, pkcs11-tool and expect
to run the tests.
```

levitte commented on test/recipes/95-test_external_pkcs11_provider_data/pkcs11-provider.sh:
```diff
@@ -0,0 +1,62 @@
+echo "------------------------------------------------------------------"
+echo "Running tests"
+echo "------------------------------------------------------------------"
+meson test -C $PKCS11_PROVIDER_BUILDDIR
```
```suggestion
# The OpenSSL app uses ${HARNESS_OSSL_PREFIX} as a prefix for its standard output
HARNESS_OSSL_PREFIX= meson test -C $PKCS11_PROVIDER_BUILDDIR
```

The-Mule replied:
Thank you, that fixed the issue.

beldmit commented:
@t8m I think @The-Mule should be covered by the CCLA

t8m commented:
@beldmit We need an ICLA from Ondrej. @The-Mule can you please submit it? https://openssl-library.org/policies/cla/index.html Please mention that you are covered by Red Hat CCLA.

beldmit commented:
@The-Mule you need to update https://github.com/openssl/openssl/blob/c477fa5a22ff27081b1725ecef21c61ae0d7a587/.github/workflows/ci.yml#L567-L603

The-Mule commented:
Thank you all for such a quick and perfectly constructive feedback. I rebased the commit to include all the changes. Also I added test execution into GH actions workflow ci.yml with all the dependencies needed, hopefully.

beldmit reviewed (dismissed):
LGTM

t8m commented on test/README-external.md:
```diff
@@ -118,6 +118,35 @@ algorithms to be skipped, e.g. as follows:
 The names of all supported quantum-safe algorithms are available at
 <https://github.com/open-quantum-safe/oqs-provider#algorithms>
 
+pkcs11-provider test suite
+======================
```
```suggestion
pkcs11-provider test suite
==========================
```

The-Mule replied:
Fixed.

t8m commented on test/README-external.md:
```diff
@@ -118,6 +118,35 @@ algorithms to be skipped, e.g. as follows:
+pkcs11-provider test suite
+======================
+
+This builds and runs pkcs11-provider tests agains the local OpenSSL build.
```
Typo: against

The-Mule replied:
Fixed.

t8m commented on test/README-external.md:
```diff
@@ -118,6 +118,35 @@ algorithms to be skipped, e.g. as follows:
+
+Then configure/build OpenSSL enabling external tests:
+
+    $ ./config shared enable-external-tests
```
Please remove shared (that is a default option)

The-Mule replied:
Would you mind me fixing that for other parts of README-external.md?

t8m replied:
Please fix it wherever applicable.

The-Mule replied:
Fixed.

t8m commented on .github/workflows/ci.yml:
```diff
@@ -602,6 +602,27 @@ jobs:
         gdb < <(echo -e "file ./libcrypto.so.3\nquit") > ./results
         grep -q "Reading symbols from.*libcrypto\.so\.3\.debug" results
 
+  external-test-pkcs11-provider:
```
I am worried about ever-expanding number of CI jobs run on PRs. Not sure why this needs a special CI job, why it cannot be integrated into another one.

beldmit replied:
Separating this one was my idea because the existing external test CI job already takes 10 minutes and this task needs its own set of specific packages. I don't insist on my proposal

t8m replied:
We could instead merge the other external CI jobs to one. They should not take more than 10 mins in total.

The-Mule replied:
I can merge pkcs11 and osq provider to external_test_providers and merge external-test-cf-quiche into external_tests (or I can rename that to external_tests_misc), that should keep all under 10 minutes while removing two separate tests (pkcs11 and cf-quiche). pyca external tests contains matrix and hence it probably should stay separated (even though it currently runs in a single configuration). So there would be:

1. external_tests(_misc) (~9 mins)
2. external_tests_providers (~6 mins)
3. external_tests_pyca (~5 mins)

t8m replied:
Sure, that WFM.

The-Mule replied:
Done (it is actaully ~8 mins, ~7 mins and ~6 mins now.

beldmit reviewed (approved):
LGTM

mattcaswell commented:
Pushed to master. Thanks for your contribution.
